kubernetes/deployment/controller.go

```diff
@@ -0,0 +1,295 @@
+	fmt.Println("====================")
+	fmt.Printf("\n\n")
+	return err
+}
+
+// Process a generic object
+func (c *Controller) processObj(obj interface{}) error {
+	fmt.Println("processObj called")
+	fmt.Printf("Object has type %T\n", obj)
+	switch v := obj.(type) {
+	case *apiv1.Pod:
+		fmt.Println("Case pod. Calling processPod")
+		c.processPod(v)
+	default:
+		fmt.Println("Default case executed, throwing error")
+		return error(fmt.Errorf("obj is not a pod"))
+	}
+	return nil
+}
+
+// processPod defines how to react to a pod coming off the workqueue
+// in an observed state.
+// Assumes there is only 1 initcontainer
+func (c *Controller) processPod(obj *apiv1.Pod) {
+	fmt.Println("processPod Called")
+	if len(obj.Status.InitContainerStatuses) != 0 {
+		switch {
+		case obj.Status.InitContainerStatuses[0].State.Waiting != nil:
+			c.Opts.Logger.Debug(fmt.Sprintf("InitContainer Waiting!"))
+
+		case obj.Status.InitContainerStatuses[0].State.Running != nil:
+			c.Opts.Logger.Debug(fmt.Sprintf("InitContainer Running!"))
+			c.Opts.Logger.Info(fmt.Sprintf("Calling CopyTar with files: %+v", c.Opts.Files))
+			err := c.Client.CopyTar(c.Opts.Files, obj)
+			if err != nil {
+				c.Opts.Logger.Error(fmt.Sprintf("Error copying tarball: %s", err))
+			}
+		case obj.Status.ContainerStatuses[0].State.Running != nil:
+			// Write the pod name, name to the resources file.
+			// This allows us to retrieve it to obtain the client.token
+			resources := &cloud.Resources{}
+			file, err := os.OpenFile(c.Opts.ResourcePath, os.O_RDONLY, 0600)
+			if err != nil {
+				c.Opts.Logger.Error(fmt.Sprintf("Could not open resource file with path: %s", err))
+			}
+			c.Opts.Logger.Info(fmt.Sprintf("Opened resource file with path %s", c.Opts.ResourcePath))
+			decoder := gob.NewDecoder(file)
+			err = decoder.Decode(resources)
+			if err != nil {
+				c.Opts.Logger.Error(fmt.Sprintf("Error decoding resource file: %s", err))
+			}
+			err = file.Close()
+			if err != nil {
+				c.Opts.Logger.Error(fmt.Sprintf("Failed to close resource file: %s", err))
+			}
+			file2, err := os.OpenFile(c.Opts.ResourcePath, os.O_WRONLY, 0600)
+			if err != nil {
+				c.Opts.Logger.Error(fmt.Sprintf("Failed to open file2 %s", err))
+			}
+			resources.Details["manager-pod"] = obj.ObjectMeta.Name
+			encoder := gob.NewEncoder(file2)
+			err = encoder.Encode(resources)
+			if err != nil {
+				c.Opts.Logger.Error(fmt.Sprintf("Failed to encode resource file: %s", err))
+			}
+			err = file2.Close()
+			if err != nil {
+				c.Opts.Logger.Error(fmt.Sprintf("Failed to close resource file2: %s", err))
+			}
+
+			if err != nil {
+				c.Opts.Logger.Error(fmt.Sprintf("Failed to close resource file %s", err))
+			}
+			c.Opts.Logger.Info(fmt.Sprintf("Stored manager pod name %s in resource file", obj.ObjectMeta.Name))
+			c.Opts.Logger.Info(fmt.Sprintf("WR manager container is running, calling PortForward with ports %v", c.Opts.RequiredPorts))
+			go c.Client.PortForward(obj, c.Opts.RequiredPorts)
```

Comment:
Error return value of `c.Client.PortForward` is not checked

Reply:
@sb10 
Not sure how to check this, I could wrap it in a go func but that would lead to it repeatedly being called on the same ports. At the moment it (should) reconnect if interrupted (and the entire controller has not died / errored out for some reason)

Reply:
Not sure what you mean. It's already called in a go routine. Putting in an anonymous func won't make any difference to control flow, but will let you capture and log the error.